presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java

```diff
@@ -89,6 +89,8 @@
     private final Map<NodeRef<OrderBy>, List<FunctionCall>> orderByWindowFunctions = new LinkedHashMap<>();
 
     private final Map<NodeRef<Join>, Expression> joins = new LinkedHashMap<>();
+    private final Map<NodeRef<Join>, JoinUsingAnalysis> joinUsing = new LinkedHashMap<>();
```

Comment:
I think we could reuse existing analysis, e.g: synthesize join expressions (like it was before) but with field references, e.g:
```
new ComparisionExpression(EQUAL,
     new FieldReference(leftField.getRelationFieldIndex())),
     new FieldReference(leftField.getRelationFieldIndex())))
```
One benefit of that approach is that we gather coercion information at the analysis time and can reuse `ExpressionAnalyzer`.

Additionally, we could reuse `Analysis#outputExpressions` to "project" join output columns. So for instance output expressions would contain:
```
new Coalesce(
  new FieldReference(joinLeftField1.getRelationFieldIndex()),
  new FieldReference(joinRightField2.getRelationFieldIndex())),
...,
new FieldReference(leftField1.getRelationFieldIndex()),
...
new FieldRerference(rightField1.getRelationFieldIndex()),
)
```
This way we can clearly separate analysis from planning.  We kind of synthesize conditions with `FieldReferences`, but in your code you still synthesize join `Identifiers` so that you can retain type. What do you think?

Reply:
Synthesizing expressions is a little like planning. I do want to keep the analysis as descriptive as possible and leave it to the planner to make all the choices regarding how a given SQL construct turns into a expression in the IR.

Reply:
> Synthesizing expressions is a little like planning

Indeed, although we are not free from synthesizing with `JoinUsingAnalysis` too (we are synthesizing output identifiers). What I like mostly about creating comparison expression with field refs is that we would:
* reuse existing join planning code
* reuse expression analyzer
* properly derive types at analysis time.

We use `FieldReference` throughout the analyzer already since it allows to pass standardized information to planner (e.g: `field_ref1, field_ref2` instead of `*`), 
Basically what I mean is to gather all join information in a unified way for planner to plan.
After all `using` is some kind of syntax sugar (e.g: similar to '*') that we can turn into descriptor during analysis.

That being said, I won't object if you decide on other approach. I wonder if error messages will be displayed correctly with `JoinUsingAnalysis` (whether `NodeLocation` is preserved).